File: Projekt_Python/Swiat.py
import os
import random
from Organizm import Organizm
from zwierzeta.Antylopa import Antylopa
from zwierzeta.CyberOwca import CyberOwca
from zwierzeta.Czlowiek import Czlowiek
from zwierzeta.Lis import Lis
from zwierzeta.Owca import Owca
from zwierzeta.Wilk import Wilk
from zwierzeta.Zolw import Zolw

from rosliny.Barszcz import Barszcz
from rosliny.Guarana import Guarana
from rosliny.Jagoda import Jagoda
from rosliny.Mlecz import Mlecz
from rosliny.Trawa import Trawa


class Swiat:

    def __init__(self, x, y):
        self.__rozmiarx = x
        self.__rozmiary = y
        self.__plansza = [[0] * x for i in range(y)]
        self.__organizmy = []
        self.__raport = []
        for i in range(0, x):
            for j in range(0, y):
                self.__plansza[i][j] = None

    def getRaport(self, i):
        return self.__raport[i]

    def setRaport(self, tekst):
        self.__raport.append(tekst)

    def setRaportClear(self):
        self.__raport.clear()

    def getRaportSize(self):
        return len(self.__raport)

    def getPlansza(self, x, y):
        return self.__plansza[x][y]

    def setPlanszaKolizja(self, x, y, organizm):
        self.__plansza[x][y] = organizm
        organizm.setWykonalRuch(True)

    def setPlansza(self, x, y, organizm):
        self.__plansza[int(x)][int(y)] = organizm
        self.__organizmy.append(organizm)

    def setPlanszaNone(self, x, y):
        self.__plansza[x][y] = None

    def getRozmiarx(self):
        return self.__rozmiarx

    def getRozmiary(self):
        return self.__rozmiary

    def getOrganizmy(self):
        return self.__organizmy

    def czyWPlanszy(self, x, y):
        if x < 0 or y < 0 or x >= self.__rozmiarx or y >= self.__rozmiary: return False
        return True

    def tworzZwierze(self, symbol, x, y):
        if symbol == 0:
            wilk = Wilk(x, y, None)
            self.setPlansza(x, y, wilk)
        elif symbol == 1:
            owca = Owca(x, y, None)
            self.setPlansza(x, y, owca)
        elif symbol == 2:
            lis = Lis(x, y, None)
            self.setPlansza(x, y, lis)
        elif symbol == 3:
            zolw = Zolw(x, y, None)
            self.setPlansza(x, y, zolw)
        elif symbol == 4:
            antylopa = Antylopa(x, y, None)
            self.setPlansza(x, y, antylopa)
        elif symbol == 6:
            czlowiek = Czlowiek(x, y, None)
            self.setPlansza(x, y, czlowiek)
        elif symbol == 5:
            cyber_owca = CyberOwca(x, y, None)
            self.setPlansza(x, y, cyber_owca)

    def tworzRosline(self, symbol, x, y):
        if symbol == 0:
            trawa = Trawa(x, y, None)
            self.setPlansza(x, y, trawa)
        elif symbol == 1:
            mlecz = Mlecz(x, y, None)
            self.setPlansza(x, y, mlecz)
        elif symbol == 2:
            guarana = Guarana(x, y, None)
            self.setPlansza(x, y, guarana)
        elif symbol == 3:
            jagoda = Jagoda(x, y, None)
            self.setPlansza(x, y, jagoda)
        elif symbol == 4:
            barszcz = Barszcz(x, y, None)
            self.setPlansza(x, y, barszcz)

    def tworzZwierzeOdczyt(self, symbol, dane):
        if symbol == 'W':
            wilk = Wilk(dane[0], dane[1], dane)
            self.setPlansza(dane[0], dane[1], wilk)
        elif symbol == 'O':
            owca = Owca(dane[0], dane[1], dane)
            self.setPlansza(dane[0], dane[1], owca)
        elif symbol == 'L':
            lis = Lis(dane[0], dane[1], dane)
            self.setPlansza(dane[0], dane[1], lis)
        elif symbol == 'Z':
            zolw = Zolw(dane[0], dane[1], dane)
            self.setPlansza(dane[0], dane[1], zolw)
        elif symbol == 'A':
            antylopa = Antylopa(dane[0], dane[1], dane)
            self.setPlansza(dane[0], dane[1], antylopa)
        elif symbol == 'Q':
            cyber_owca = CyberOwca(dane[0], dane[1], dane)
            self.setPlansza(dane[0], dane[1], cyber_owca)
        elif symbol == 'C':
            czlowiek = Czlowiek(dane[0], dane[1], dane)
            self.setPlansza(dane[0], dane[1], czlowiek)

    def tworzRoslineOdczyt(self, symbol, dane):
        if symbol == 't':
            trawa = Trawa(dane[0], dane[1], dane)
            self.setPlansza(dane[0], dane[1], trawa)
        elif symbol == 'm':
            mlecz = Mlecz(dane[0], dane[1], dane)
            self.setPlansza(dane[0], dane[1], mlecz)
        elif symbol == 'g':
            guarana = Guarana(dane[0], dane[1], dane)
            self.setPlansza(dane[0], dane[1], guarana)
        elif symbol == 'j':
            jagoda = Jagoda(dane[0], dane[1], dane)
            self.setPlansza(dane[0], dane[1], jagoda)
        elif symbol == 'b':
            barszcz = Barszcz(dane[0], dane[1], dane)
            self.setPlansza(dane[0], dane[1], barszcz)

    def wyoknajTure(self):
        k = 0
        self.__organizmy.sort(key=Organizm.getInicjatywa)
        for j in range(0, len(self.__organizmy)):
            if j >= len(self.__organizmy): break
            tmp = self.__organizmy[j]
            if tmp.getZycie() == 0: break
            if tmp._wykonal_ruch == True: continue
            tmp.setWykonalRuch(True)
            xprev = tmp.getX()
            yprev = tmp.getY()
            tmp.setXprev(xprev)
            tmp.setYprev(yprev)
            self.setPlanszaNone(tmp.getX(),tmp.getY())

            tmp.wykonajRuch(self, self.__organizmy, j)
            if self.__plansza[tmp.getX()][tmp.getY()] != None:
                obronca = self.__plansza[tmp.getX()][tmp.getY()]
                obronca.kolizja(tmp, self.__organizmy, j, self)
            else:
                self.__plansza[tmp.getX()][tmp.getY()] = tmp

        for j in range(0, len(self.__organizmy)):
            self.__organizmy[j].setWykonalRuch(False)
            self.__organizmy[j].setZycie(self.__organizmy[j].getZycie() + 1)


    def generujSwiat(self):  # ,okno):
        while True:
            x = random.randint(0, self.getRozmiarx() - 1)
            y = random.randint(0, self.getRozmiary() - 1)
            if self.getPlansza(x, y) == None: break

        self.tworzZwierze(6, x, y)

        for i in range(0, self.getRozmiarx() - 1):
            organizm = random.randint(0, 5)
            while True:
                x = random.randint(0, self.getRozmiarx() - 1)
                y = random.randint(0, self.getRozmiary() - 1)
                if self.getPlansza(x, y) == None: break
            self.tworzZwierze(organizm, x, y)

        for i in range(0, self.getRozmiary() - 1):
            organizm = random.randint(0, 5)
            while True:
                x = random.randint(0, self.getRozmiarx() - 1)
                y = random.randint(0, self.getRozmiary() - 1)
                if self.getPlansza(x, y) == None: break
            self.tworzRosline(organizm, x, y)

        self.wyoknajTure()

    def zapisDoPliku(self):
        plik = open("zapis.txt", "w+")
        # plik.write("20 " + " " + "20\n")

        for i in range(0, len(self.__organizmy)):
            organizm = self.__organizmy[i]
            plik.write(str(organizm) + '\n')

        plik.close()

    def odczytZPliku(self):
        self.__organizmy.clear()
        for i in range(0,20):
            for j in range(0,20):
                self.setPlanszaNone(i,j)
        plik = open("zapis.txt", "r")
        for line in plik:
            dane = line.split(',')
            znak = dane[0][1]
            del dane[0]
            for i in range(0,len(dane)):
                dane[i]=int(dane[i])

            if znak >= 'A' and znak <= 'Z':
                self.tworzZwierzeOdczyt(znak, dane)
            else:
                self.tworzRoslineOdczyt(znak, dane)

File: projekt_java/src/po/java/Swiat.java
package po.java;

import po.java.rosliny.*;
import po.java.zwierzeta.*;

import java.awt.*;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.util.*;

public class Swiat {
    private
    Organizm[][] plansza;
    ArrayList<Organizm> organizmy= new ArrayList<Organizm>();



    public ArrayList <String> raport= new ArrayList<String>();

    public String getRaport(int i) {
        return raport.get(i);
    }

    public void setRaport(String tekst) {
        this.raport.add(tekst);
    }
    public void setRaportClear() {
        this.raport.clear();
    }
    public int getRaportSize() {
        return raport.size();
    }

    int rozmiarx,rozmiary;
    public Swiat(int x,int y)
    {
        this.rozmiarx=x;
        this.rozmiary=y;
        this.plansza=new Organizm[x][y];
        for(int i=0; i<rozmiarx; i++) {
            for (int j = 0; j < rozmiary; j++) {
                plansza[i][j]=null;
            }
        }
    }

    public Organizm getPlansza(int x, int y) {
        return plansza[x][y];
    }
    public void setPlanszaKolizja(int x, int y, Organizm organizm)
    {
        this.plansza[x][y] = organizm;
        organizm.setWykonalRuch(true);
    }
    public void setPlansza(int x,int y, Organizm organizm) {
        this.plansza[x][y] = organizm;
        this.organizmy.add(organizm);
    }
    public void setPlansza(int x, int y)
    {
        this.plansza[x][y]=null;
    }

    public int getRozmiarx() {
        return rozmiarx;
    }

    public int getRozmiary() {
        return rozmiary;
    }

    public ArrayList<Organizm> getOrganizmy() {
        return organizmy;
    }

    public boolean czyWPlanszy(Swiat swiat, int x, int y) {
        if (x < 0 || y < 0 || x >=rozmiarx  || y >= rozmiary) return false;
        return true;
    }
    public void tworzZwierze(int symbol , int x, int y)
    {
        if (symbol == 0)
        {
            Wilk wilk = new Wilk(x, y);
            setPlansza(x, y, wilk);
        }
        else if (symbol == 1)
        {
            Owca owca = new Owca(x, y);
            setPlansza(x, y, owca);
        }
        else if (symbol == 2)
        {
            Lis lis = new Lis(x, y);
            setPlansza(x, y, lis);
        }
        else if (symbol == 3)
        {
            Zolw zolw = new Zolw(x, y);
            setPlansza(x, y, zolw);
        }
        else if (symbol == 4)
        {
            Antylopa antylopa = new Antylopa(x, y);
            setPlansza(x, y, antylopa);
        }
        else if (symbol == 5)
        {
            Czlowiek czlowiek = new Czlowiek(x, y);
            setPlansza(x, y, czlowiek);
        }
    }
    public void tworzRosline(int symbol , int x, int y)
    {
        if (symbol == 0)
        {
            Trawa trawa = new Trawa(x, y);
            setPlansza(x, y, trawa);
        }
        else if (symbol == 1)
        {
            Mlecz mlecz = new Mlecz(x, y);
            setPlansza(x, y, mlecz);
        }
        else if (symbol == 2)
        {
            Guarana guarana = new Guarana(x, y);
            setPlansza(x, y, guarana);
        }
        else if (symbol == 3)
        {
            Jagoda jagoda = new Jagoda(x, y);
            setPlansza(x, y, jagoda);
        }
        else if (symbol == 4)
        {
            Barszcz barszcz = new Barszcz(x, y);
            setPlansza(x, y, barszcz);
        }
    }
    public void tworzZwierzeOdczyt(char symbol , int []dane)
    {
        if (symbol == 'W')
        {
            Wilk wilk = new Wilk(dane);
            setPlansza(dane[0], dane[1], wilk);
        }
        else if (symbol == 'O')
        {
            Owca owca = new Owca(dane);
            setPlansza(dane[0], dane[1], owca);
        }
        else if (symbol == 'L')
        {
            Lis lis = new Lis(dane);
            setPlansza(dane[0], dane[1], lis);
        }
        else if (symbol == 'Z')
        {
            Zolw zolw = new Zolw(dane);
            setPlansza(dane[0], dane[1], zolw);
        }
        else if (symbol == 'A')
        {
            Antylopa antylopa = new Antylopa(dane);
            setPlansza(dane[0], dane[1], antylopa);
        }
        else if (symbol == 'C')
        {
            Czlowiek czlowiek = new Czlowiek(dane);
            setPlansza(dane[0], dane[1], czlowiek);
        }
    }
    public void tworzRoslineOdczyt(char symbol , int []dane)
    {
        if (symbol == 't')
        {
            Trawa trawa = new Trawa(dane);
            setPlansza(dane[0], dane[1], trawa);
        }
        else if (symbol == 'm')
        {
            Mlecz mlecz = new Mlecz(dane);
            setPlansza(dane[0], dane[1], mlecz);
        }
        else if (symbol == 'g')
        {
            Guarana guarana = new Guarana(dane);
            setPlansza(dane[0], dane[1], guarana);
        }
        else if (symbol == 'j')
        {
            Jagoda jagoda = new Jagoda(dane);
            setPlansza(dane[0], dane[1], jagoda);
        }
        else if (symbol =='b')
        {
            Barszcz barszcz = new Barszcz(dane);
            setPlansza(dane[0], dane[1], barszcz);
        }
    }
    public void wyoknajTure()
    {
        int k=0,xprev,yprev;
        Organizm tmp,obronca;

        Comparator<Organizm> comparator= (o1,o2)->
        {
            if(o1.getInicjatywa()==o2.getInicjatywa())
            {
                return o2.getZycie()-o1.getZycie();
            }
            else return o2.getInicjatywa()-o1.getInicjatywa();
        };
        Collections.sort(organizmy,comparator);

        for( int j=0; j<organizmy.size(); j++)
        {
            tmp = organizmy.get(j);
            if(tmp.getZycie()==0) break;
            if(tmp.wykonal_ruch==true) continue;
            tmp.setWykonalRuch(true);
            xprev = tmp.getX();
            yprev = tmp.getY();
            tmp.setXprev(xprev);
            tmp.setYprev(yprev);
            plansza[tmp.getX()][tmp.getY()] =null;

            tmp.wykonajRuch(this,organizmy,j);

            if (plansza[tmp.getX()][tmp.getY()] != null)
            {
                obronca = plansza[tmp.getX()][tmp.getY()];
                obronca.kolizja(tmp, organizmy, j, this);
            }
            else
                plansza[tmp.getX()][tmp.getY()] = tmp;
        }
        for (int j = 0; j < organizmy.size(); j++)
        {
            organizmy.get(j).setWykonalRuch(false);
            organizmy.get(j).setZycie(organizmy.get(j).getZycie() + 1);
        }
    }

    public void generujSwiat(Okienko okno)
    {
        Random liczba=new Random();
        int x,y,organizm;
        do {
            x = liczba.nextInt(getRozmiarx());
            y = liczba.nextInt(getRozmiary());
        } while(getPlansza(x,y)!=null);
        tworzZwierze(5, x, y);
        for(int i=0; i<getRozmiarx()-1; i++)
        {
            organizm=liczba.nextInt(5);
            do {
                x = liczba.nextInt(getRozmiarx());
                y = liczba.nextInt(getRozmiary());
            } while(getPlansza(x,y)!=null);
            tworzZwierze(organizm,x,y);
        }
        for(int i=0; i<getRozmiary()-1; i++)
        {
            organizm=liczba.nextInt(5);
            do {
                x = liczba.nextInt(getRozmiarx());
                y = liczba.nextInt(getRozmiary());
            } while(getPlansza(x,y)!=null);
            tworzRosline(organizm,x,y);
        }
        okno.createOkienko(this);
        wyoknajTure();
    }
    public void zapisDoPliku() throws FileNotFoundException {
        PrintWriter zapis= new PrintWriter("organizmy.txt");
        zapis.println(rozmiarx+" "+rozmiary);
        for(Organizm organizm: organizmy)
        {
            zapis.println(organizm.toString());
        }
        zapis.close();
    }
    public void odczytZPliku(File file, Scanner in) throws FileNotFoundException {
        String zdanie,wyrazy[];
        char znak;
        organizmy.clear();
        while(in.hasNextLine())
        {
            zdanie=in.nextLine();
            wyrazy=zdanie.split(" ");
            znak=wyrazy[0].charAt(0);

            int []dane= new int[ wyrazy.length-1];

            for(int i=1; i<wyrazy.length; i++)
            {
                dane[i-1]=Integer.parseInt(wyrazy[i]);
            }

            if(znak>='A'&&znak<='Z')
                tworzZwierzeOdczyt(znak,dane);
            else
                tworzRoslineOdczyt(znak,dane);
        }
    }
}
